drivers/ethernet/eth_esp32.c

```diff
@@ -0,0 +1,284 @@
+/*
+ * Copyright (c) 2021 Espressif Systems (Shanghai) Co., Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT espressif_esp32_eth
+
+#include <ethernet/eth_stats.h>
+#include <zephyr/drivers/interrupt_controller/intc_esp32.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/net/ethernet.h>
+
+#include <driver/periph_ctrl.h>
+#include <esp_attr.h>
+#include <esp_mac.h>
+#include <hal/emac_hal.h>
+#include <hal/emac_ll.h>
+
+#include "eth.h"
+
+LOG_MODULE_REGISTER(eth_eth_esp32, CONFIG_ETHERNET_LOG_LEVEL);
+
+#define MAC_RESET_TIMEOUT_MS 100
+
+struct eth_esp32_dma_data {
+	uint8_t descriptors[
+		CONFIG_ETH_DMA_RX_BUFFER_NUM * sizeof(eth_dma_rx_descriptor_t) +
+		CONFIG_ETH_DMA_TX_BUFFER_NUM * sizeof(eth_dma_tx_descriptor_t)];
+	uint8_t rx_buf[CONFIG_ETH_DMA_RX_BUFFER_NUM][CONFIG_ETH_DMA_BUFFER_SIZE];
+	uint8_t tx_buf[CONFIG_ETH_DMA_TX_BUFFER_NUM][CONFIG_ETH_DMA_BUFFER_SIZE];
+};
+
+struct eth_esp32_dev_data {
+	struct net_if *iface;
+	uint8_t mac_addr[6];
+	emac_hal_context_t hal;
+	struct eth_esp32_dma_data *dma;
+	uint8_t txb[NET_ETH_MAX_FRAME_SIZE];
+	uint8_t rxb[NET_ETH_MAX_FRAME_SIZE];
+	uint8_t *dma_rx_buf[CONFIG_ETH_DMA_RX_BUFFER_NUM];
+	uint8_t *dma_tx_buf[CONFIG_ETH_DMA_TX_BUFFER_NUM];
+	struct k_sem int_sem;
+	K_KERNEL_STACK_MEMBER(rx_thread_stack, CONFIG_ETH_ESP32_RX_THREAD_STACK_SIZE);
+	struct k_thread rx_thread;
+};
+
+struct eth_esp32_dev_config {
+	const char *phy_connection_type;
+};
+
+static enum ethernet_hw_caps eth_esp32_caps(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+	return ETHERNET_LINK_10BASE_T | ETHERNET_LINK_100BASE_T;
+}
+
+static int eth_esp32_send(const struct device *dev, struct net_pkt *pkt)
+{
+	struct eth_esp32_dev_data *dev_data = dev->data;
+	size_t len = net_pkt_get_len(pkt);
+
+	if (net_pkt_read(pkt, dev_data->txb, len)) {
+		return -EIO;
+	}
+
+	uint32_t sent_len = emac_hal_transmit_frame(&dev_data->hal, dev_data->txb, len);
+
+	int res = len == sent_len ? 0 : -EIO;
+	return res;
+}
+
+static struct net_pkt *eth_esp32_rx(
+	struct eth_esp32_dev_data *const dev_data, uint32_t *frames_remaining)
+{
+	uint32_t free_rx_descriptor;
+	uint32_t receive_len = emac_hal_receive_frame(
+		&dev_data->hal, dev_data->rxb, sizeof(dev_data->rxb),
+		frames_remaining, &free_rx_descriptor);
+	if (receive_len == 0) {
+		// Nothing to receive
+		return NULL;
+	}
+
+	struct net_pkt *pkt = net_pkt_rx_alloc_with_buffer(
+		dev_data->iface, receive_len, AF_UNSPEC, 0, K_NO_WAIT);
```

Comment:
I might change this timeout from `K_NO_WAIT` to `K_MSEC(100)` to allow time for a network packet to become free.

There is also software flow control possible for RX DMA buffers that is not implemented here yet (`CONFIG_ETH_SOFT_FLOW_CONTROL`)